Use .godot as file extension for project files.

27thLiz opened the pull request:
Now project files don't have to be named "godot.cfg" anymore, they can have any name so as long as it ends with *.godot.
Also godot will automatically start the editor now if launched with a project file as an argument.
This allows for double-clicking of projects to open them :)

Code-wise this should be complete, but there's still work to do:

- Make a nice icon for godot projects. (ping @djrm)
- Work on installers/packaging -> register the extension and icon with godot. (ping @Calinou)
- Update the 2.1 to 3.0 exporter.

Tested on linux and windows so far.
Closes #6915

akien-mga commented on core/global_config.cpp:
```diff
@@ -236,13 +241,43 @@ bool GlobalConfig::_load_resource_pack(const String &p_pack) {
+	String ret = "";
+	while (true) {
+		bool isdir;
+		String f = p_da->get_next(&isdir);
```
I know this is already done this way in various other places, but I think we should start to enforce using slightly more explicit variable names (i.e. no one-char names apart from iterations). Here `file` would be good IMO, maybe even `p_diraccess` too. WDYT?

Hinsbart replied:
sounds good

Hinsbart replied:
updated, no more single character variables.

akien-mga commented on core/global_config.cpp:
```diff
@@ -236,13 +241,43 @@ bool GlobalConfig::_load_resource_pack(const String &p_pack) {
-
-		if (_load_settings("res://godot.cfg") == OK || _load_settings_binary("res://godot.cfb") == OK) {
+		String gdproj = _find_project_file();
+		if (_load_settings(gdproj) == OK || _load_settings_binary("res://godot.cfb") == OK) {
```
What should we do with the `godot.cfb`? Does it really serve a purpose, or should we deprecate it? If we keep it, should we make it use a similar system, e.g. `<name>.godotb` (ugh)?

Hinsbart replied:
Well, good question. I don't know what kind of performance difference a binary config file makes, I guess @reduz should clear this up :P 
The reason I kept it at `godot.cfb` was that usually neither the game developer nor the player will ever care about this file, so the name doesn't really matter IMO.

vnen commented:
🎉 Awesome!

27thLiz committed:
Use .godot as file extension for project files.

Now project files don't have to be named "godot.cfg" anymore, they can have any name so as long as it ends with *.godot.
Also godot will automatically start the editor now if launched with a project file as an argument.
This allows for double-clicking of projects to open them :)

Code-wise this should be complete, but there's still work to do:

- Make a nice icon for godot projects.
- Work on installers/packaging -> register the extension and icon with godot.
- Update the 2.1 to 3.0 exporter.

Tested on linux and windows so far.

reduz commented:
Why was this merged? This feature is 100% unnecesary.

reduz commented:
If you want to associate to a file explorer from the OS, you can change config name to something like config.godot, but having custom naming is more trouble and completely unnecesary.

reduz commented:
Also it leaves for the confusing situation where having multiple .godot files in the project folder makes no sense anywah

reduz commented:
I think changing extension to .godot or something like config.godot, and making godot open the editor for those, as well as associating the extension to godot in windows and other OSs is an excellent idea (i did not get the use for that in the begining, so apologies), I'm just against the custom naming the project file part, which I believe is unnecesary... If you want to redo the PR sans that, I'm fine.

jmf commented:
Many engines have their own project file format, and clicking a file to launch that project in the engine is easier than launching godot, looking through a long list of projects and finding it somewhere. When not using any VCS for collaborating and sending zipped projects, it also makes it easier to just launch the project without having to import it first. The extension ".cfg" is just too common to be a good extension to use with MIME types. I don't want my syslinux.cfg be opened with godot, but opening every <name>.godot with godot makes sense.

volzhs commented:
@reduz I agreed. just `config.godot` would be good enough.
custom name like `my_project.godot` `my_tutorial.godot` seems not necessary.

reduz commented:
@jmf yeah.. I just didn't realize that use case as first, so the feature seemed useless. That said, everything else than custom file naming is fine.

volzhs commented:
@reduz by the way... what about "recent files" on OS?
there would be multiple "config.godot" file list if someone works on many projects at the same time.
I agreed this original PR now. hm...

reduz commented:
projects have names, so i suppose that is enough?
otherwise you can take the path the project is on (rightmost folder)

On Sat, Apr 29, 2017 at 6:11 PM volzhs <notifications@github.com> wrote:

> @reduz <https://github.com/reduz> by the way... what about "recent files"
> on OS?
> there would be multiple "config.godot" file list if someone works on many
> projects at the same time.
> I agreed this original PR now. hm...
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/8375#issuecomment-298178099>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AF-Z28JCoTI4dKyK05Oo7uiHAOj5C8gZks5r02EfgaJpZM4M7U2Y>
> .
>

volzhs commented:
@reduz I meant OS feature, like this.
![image](https://cloud.githubusercontent.com/assets/8281454/25556995/02e96bae-2d43-11e7-90a4-21cbc243d3b2.png)

reduz commented:
ah, you mean recent documents.. I see, I still dont think it should apply
because a godot project is not a document, and most users open godot from
the project manager anyway.

On Sat, Apr 29, 2017 at 6:19 PM volzhs <notifications@github.com> wrote:

> @reduz <https://github.com/reduz> I meant OS feature, like this.
> [image: image]
> <https://cloud.githubusercontent.com/assets/8281454/25556995/02e96bae-2d43-11e7-90a4-21cbc243d3b2.png>
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/8375#issuecomment-298178519>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AF-Z25n-XF7H1_4QvhGhEFO9z0VW49JBks5r02MNgaJpZM4M7U2Y>
> .
>

volzhs commented:
@reduz I agreed with you partially, and I think I appealed what I think about this. :)
any other thought on this?

reduz commented:
think of it this way.. forcing users to name their project so it appears in
the recent files is too much..
maybe instead of doing a rename to config.godot , we can leave godot.cfg as
is and rename scene files to gscn/tscn and register those in the shell?

On Sat, Apr 29, 2017 at 6:37 PM Juan Linietsky <reduzio@gmail.com> wrote:

> ah, you mean recent documents.. I see, I still dont think it should apply
> because a godot project is not a document, and most users open godot from
> the project manager anyway.
>
> On Sat, Apr 29, 2017 at 6:19 PM volzhs <notifications@github.com> wrote:
>
>> @reduz <https://github.com/reduz> I meant OS feature, like this.
>> [image: image]
>> <https://cloud.githubusercontent.com/assets/8281454/25556995/02e96bae-2d43-11e7-90a4-21cbc243d3b2.png>
>>
>> —
>> You are receiving this because you were mentioned.
>>
>>
>> Reply to this email directly, view it on GitHub
>> <https://github.com/godotengine/godot/pull/8375#issuecomment-298178519>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AF-Z25n-XF7H1_4QvhGhEFO9z0VW49JBks5r02MNgaJpZM4M7U2Y>
>> .
>>
>

volzhs commented:
@reduz opening a tscn can open project?

reduz commented:
if we register the shell command to be executed as godot -e, then yes

On Sat, Apr 29, 2017 at 6:47 PM volzhs <notifications@github.com> wrote:

> @reduz <https://github.com/reduz> opening a tscn can open project?
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/8375#issuecomment-298180063>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AF-Z227amer3ocBsIQqvYdZb6D7DfqkQks5r02mhgaJpZM4M7U2Y>
> .
>

reduz commented:
imo opening single scene files is probably more useful than opening a
project
the only thing it has to be added is that the same instance of godot
editing the project has to open all the scenes..

On Sat, Apr 29, 2017 at 6:49 PM Juan Linietsky <reduzio@gmail.com> wrote:

> if we register the shell command to be executed as godot -e, then yes
>
> On Sat, Apr 29, 2017 at 6:47 PM volzhs <notifications@github.com> wrote:
>
>> @reduz <https://github.com/reduz> opening a tscn can open project?
>>
>> —
>> You are receiving this because you were mentioned.
>>
>>
>> Reply to this email directly, view it on GitHub
>> <https://github.com/godotengine/godot/pull/8375#issuecomment-298180063>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AF-Z227amer3ocBsIQqvYdZb6D7DfqkQks5r02mhgaJpZM4M7U2Y>
>> .
>>
>

volzhs commented:
then sounds cool.

eon-s commented:
But I like to open scene files with a text editor :sob: 
If going to register scene types, make this optional, please.

vnen commented:
@eon-s this is handled by the OS, you can always change it to open in any program you like.

eon-s commented:
@vnen I know, but if auto registers may be a problem, more for people that have many versions on the same system.

punto- commented:
On certain platforms (for example Steam), you'll have multiple executables
named differently for the same project, that's why naming the project file
after the executable is not an idea I like. You might also have multiple
executables in your system that you use to open the same project to test
different things. And it's nice to just create an empty engine.cfg (or
whatever name) on a directory and have that become a project.

On 29 April 2017 at 16:01, eon-s <notifications@github.com> wrote:

> @vnen <https://github.com/vnen> I know, but if auto registers may be a
> problem, more for people that have many versions on the same system.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/8375#issuecomment-298187969>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGVmPQOerC7rSzKuPOYvViB61IMHz7Lqks5r04jugaJpZM4M7U2Y>
> .
>

reduz commented:
On windows you can choose the default app to open it, so it should not be a
problem

On Apr 29, 2017 8:04 PM, "eon-s" <notifications@github.com> wrote:

> But I like to open scene files with a text editor 😭
> If going to register scene types, make this optional, please.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/8375#issuecomment-298184675>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AF-Z2-2rCVeVhWaAjKYb9IfnEk6gQlGCks5r03uPgaJpZM4M7U2Y>
> .
>
